<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="IntroPhylogenetics">
<title>Building Trees • IntroPhylogenetics</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Building Trees">
<meta property="og:description" content="IntroPhylogenetics">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">IntroPhylogenetics</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="https://www.ahl27.com/IntroPhylogenetics/articles/IntroPhylogenetics.html">Overview</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-lessons">Lessons</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-lessons">
    <a class="dropdown-item" href="https://www.ahl27.com/IntroPhylogenetics/articles/BuildingTrees.html">Building Trees</a>
    <a class="dropdown-item" href="https://www.ahl27.com/IntroPhylogenetics/articles/ComparingTrees.html">Comparing Trees</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://www.ahl27.com/tutorials">Back to Website</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/ahl27/IntroPhylogenetics" aria-label="GitHub">
    <span class="fa fa-github"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://www.ahl27.com" aria-label="My Website">
    <span class="fa fa-potato"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Building Trees</h1>
                        <h4 data-toc-skip class="author">Aidan
Lakshman<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="mailto:ahl27@pitt.edu" class="email"&gt;ahl27@pitt.edu&lt;/a&gt;&lt;/p&gt;'><sup>1</sup></a>
</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/ahl27/IntroPhylogenetics/blob/HEAD/vignettes/BuildingTrees.Rmd" class="external-link"><code>vignettes/BuildingTrees.Rmd</code></a></small>
      <div class="d-none name"><code>BuildingTrees.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Package imports</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://ape-package.ird.fr/" class="external-link">ape</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">DECIPHER</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Helper plotting function</span></span>
<span><span class="va">plot_tree_unrooted</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">dend</span>, <span class="va">title</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">tf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/DECIPHER/man/WriteDendrogram.html" class="external-link">WriteDendrogram</a></span><span class="op">(</span><span class="va">dend</span>, file<span class="op">=</span><span class="va">tf</span>, quoteLabels<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span>  <span class="va">predTree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span><span class="va">tf</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">predTree</span>, <span class="st">'unrooted'</span>, main<span class="op">=</span><span class="va">title</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="section level2">
<h2 id="distance-matrices">Distance Matrices<a class="anchor" aria-label="anchor" href="#distance-matrices"></a>
</h2>
<p>Most phylogenetic reconstruction algorithms rely upon constructing
distance matrices. There are a variety of way to do this, but for
purposes of illustration we’ll just be using Hamming Distance, which is
one of the simplest methods. More complex methods can use nucleotide or
amino acid substitution models.</p>
<p>Let’s construct a toy dataset for purposes of illustration:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sequenceSet</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html" class="external-link">DNAStringSet</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'AGACT'</span>,</span>
<span>                              <span class="st">'AGACG'</span>,</span>
<span>                              <span class="st">'TCATT'</span>,</span>
<span>                              <span class="st">'TGCTG'</span>,</span>
<span>                              <span class="st">'AGCTG'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sequenceSet</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span></span></code></pre></div>
<p><img src="toydataset.png"></p>
<p>For each pair of sequences, we count the proportion of residues that
do not match and mark this as the distance between them. As an example
of calculating distance, consider the first two sequences in this
set:</p>
<p><span class="math display">\[\begin{align*}
&amp;1)\;\;&amp;AGACT&amp;\\
&amp;2)\;\;&amp;AGACG&amp;
\end{align*}\]</span></p>
<p>Since these sequences differ at 1 out of 5 total residues, their
distance is <span class="math inline">\(\frac{1}{5}=0.2\)</span>.</p>
<p>We can quickly calculate a complete distance matrix between these
sequences with the following:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DECIPHER/man/DistanceMatrix.html" class="external-link">DistanceMatrix</a></span><span class="op">(</span><span class="va">sequenceSet</span>, type<span class="op">=</span><span class="st">'dist'</span>, verbose<span class="op">=</span><span class="cn">F</span><span class="op">)</span></span>
<span><span class="va">dm</span></span></code></pre></div>
<pre><code><span><span class="co">##     1   2   3   4   5</span></span>
<span><span class="co">## 1 0.0 0.2 0.6 0.8 0.6</span></span>
<span><span class="co">## 2 0.2 0.0 0.8 0.6 0.4</span></span>
<span><span class="co">## 3 0.6 0.8 0.0 0.6 0.8</span></span>
<span><span class="co">## 4 0.8 0.6 0.6 0.0 0.2</span></span>
<span><span class="co">## 5 0.6 0.4 0.8 0.2 0.0</span></span></code></pre>
<p>Notice that the distance between 1 and 2 is 0.2, just as we
calculated before. Now we can begin to create trees.</p>
<hr>
</div>
<div class="section level2">
<h2 id="ultrametric-trees">Ultrametric Trees<a class="anchor" aria-label="anchor" href="#ultrametric-trees"></a>
</h2>
<div class="section level3">
<h3 id="upgma">UPGMA<a class="anchor" aria-label="anchor" href="#upgma"></a>
</h3>
<p>The simplest tree building method is the Unweighted Pair Group Method
with Arithmetic Means (UPGMA). This is a hierarchical clustering
algorithm that sequentially builds a tree from the bottom up. The
algorithm begins with a distance matrix, and proceeds in the following
steps:</p>
<ol style="list-style-type: decimal">
<li><p>Identify the pair of nodes with the lowest distance <span class="math inline">\(d_{min}\)</span>. Here I’ve multiplied the
distances by 10 to make it easier to read (it doesn’t make a difference
in the final result).</p></li>
<li><p>Create a new node <span class="math inline">\(N_{new}\)</span>
that joins these pairs of nodes, and create branch lengths such that
each leaf is equidistant from <span class="math inline">\(N_{new}\)</span> The total branch length should
from the <span class="math inline">\(N_{new}\)</span> to any leaf should
be half of <span class="math inline">\(d_{min}\)</span>. In this case,
<span class="math inline">\(d_{min} = 2\)</span> and our new node is
called <span class="math inline">\(u\)</span>.</p></li>
<li><p>Combine the two rows of the distance matrix into one row by
replacing each pair of entries with their average. The average is
weighted by the number of leaves in each pair being combined, so if the
two rows are each clusters with multiple leaves, the average should be
weighted accordingly.</p></li>
</ol>
<p><img src="upgma1.png"></p>
<ol start="4" style="list-style-type: decimal">
<li>Repeat above steps until all nodes are combined and the distance
matrix is a single row and column.</li>
</ol>
<p><img src="upgma4.png"></p>
<p><img src="upgma5.png"></p>
<p><img src="upgma6.png"></p>
<p>Note that UPGMA methods always create an ultrametric tree, which
means the tree is rooted and all leaves are equidistant from the root.
This implies a molecular clock (rate of mutation is the same across all
lineages of a tree).</p>
</div>
<div class="section level3">
<h3 id="wpgma">WPGMA<a class="anchor" aria-label="anchor" href="#wpgma"></a>
</h3>
<p>A variation of UPGMA is WPGMA, where W stands for ‘weighted’. The
difference between UPGMA and WPGMA is a bit counter-intuitive. Recall
that when we combine two rows of the distance matrix in UPGMA, we weight
our average by the number of leaves in the node being combined. In
WPGMA, we do <strong>not</strong> weight by number of leaves in the
node, and instead treat each node as having equal weight.</p>
<p>As an example, consider the following distance matrix:</p>
<p><span class="math display">\[ \begin{matrix}
    &amp; n_1 &amp; n_2 &amp; n_3 \\
n_1 &amp; 0   &amp; 3   &amp; 5  \\
n_2 &amp; -   &amp; 0   &amp; 3  \\
n_3 &amp; -   &amp; -   &amp; 0  \\
\end{matrix}\]</span></p>
<p>Now suppose that node <span class="math inline">\(n_1\)</span> has 10
leaves, node <span class="math inline">\(n_2\)</span> has 2 leaves, and
node <span class="math inline">\(n_3\)</span> has 1 leaf. If we wanted
to combine nodes <span class="math inline">\(n_1\)</span> and <span class="math inline">\(n_2\)</span> into node <span class="math inline">\(u\)</span> using UPGMA, the distance from <span class="math inline">\(u\)</span> to <span class="math inline">\(n_3\)</span> would be calculated with the
following formula:</p>
<p><span class="math display">\[\begin{align*}
d_{UPGMA}(u, n_3) &amp;= \frac{1}{|n_1| + |n_2|} \left(|n_1|*d(n_1,n_3)
+ |n_2|*d(n_2,n_3) \right)\\
\\
&amp;= \frac{1}{10+2} \left(10*5 + 2*3 \right) \\
\\
&amp;= 4.67
\end{align*}\]</span></p>
<p>However, if we combined these with WPGMA, we would instead ignore the
number of leaves in each cluster and just take the simple arithmetic
average, as follows:</p>
<p><span class="math display">\[\begin{align*}
d_{WPGMA}(u, n_3) &amp;= \frac{1}{2} \left(d(n_1,n_3) + d(n_2,n_3)
\right)\\
\\
&amp;= \frac{1}{2} \left(5 + 3 \right) \\
\\
&amp;= 4
\end{align*}\]</span></p>
<p>You’re probably wondering, “why does <em>weighted</em> PGMA use
<em>unweighted</em> combinations and vice-versa?” The answer lies in
what we’re referring to when we talk about pair combinations being
(un)weighted. Since UPGMA weights each node by the number of leaves,
it’s correcting for nodes representing clades of difference sizes. In
other words, the combination is inherently weighted, and UPGMA uses a
clever way to “undo” this combination. WPGMA, by contrast, does not
weight by clade size, and thus the combinations are biased (weighted) by
the number of leaves in each nodes.</p>
</div>
<div class="section level3">
<h3 id="r-implementation">R Implementation<a class="anchor" aria-label="anchor" href="#r-implementation"></a>
</h3>
<p>We can create a UPGMA tree in R with the below commands. Note that R
doubles the lengths obtained by this, so I’m going to multiply the
distance matrix by 5 to get the same scale factor as the example worked
through above.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># UPGMA Tree</span></span>
<span><span class="va">dend</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dendrogram.html" class="external-link">as.dendrogram</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/hclust.html" class="external-link">hclust</a></span><span class="op">(</span><span class="va">dm</span><span class="op">*</span><span class="fl">5</span>, method<span class="op">=</span><span class="st">'average'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">dend</span><span class="op">)</span></span></code></pre></div>
<p><img src="BuildingTrees_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<p>Implementing WPGMA trees is very similar to UPGMA trees (and gives
the same result in this example):</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># WPGMA Tree</span></span>
<span><span class="va">dend</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dendrogram.html" class="external-link">as.dendrogram</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/hclust.html" class="external-link">hclust</a></span><span class="op">(</span><span class="va">dm</span><span class="op">*</span><span class="fl">5</span>, method<span class="op">=</span><span class="st">'mcquitty'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">dend</span><span class="op">)</span></span></code></pre></div>
<p><img src="BuildingTrees_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<p>There are also other methods for combining the distance matrices–see
the help page for <code>hclust</code> for other examples.</p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="neighbor-joining-trees">Neighbor Joining Trees<a class="anchor" aria-label="anchor" href="#neighbor-joining-trees"></a>
</h2>
<p>Another common approach to building trees is the Neighbor-Joining
(NJ) method. This method also utilizes distance matrices, but proceeds
from the top-down rather than UPGMA’s bottom-up approach. We’ll use the
same distance matrix (multiplied by 10 as before for simpler
visualization).</p>
<div class="section level3">
<h3 id="algorithm">Algorithm<a class="anchor" aria-label="anchor" href="#algorithm"></a>
</h3>
<p>The algorithm proceeds in the following steps:</p>
<ol style="list-style-type: decimal">
<li>Start with a star tree (all connected to a single root node), and
record the row and column sums for each row/column of the matrix.</li>
</ol>
<p><img src="nj1.png"></p>
<ol start="2" style="list-style-type: decimal">
<li>Create a Q-matrix, which is a matrix of the same dimensions as the
distance matrix, and each element given by the following formula:</li>
</ol>
<p><span class="math display">\[Q[i,j] = (n-2)d(i,j) - R(i) -
C(j)\]</span></p>
<div class="line-block">      Here <span class="math inline">\(n\)</span> is the number of nodes under
consideration (the number of<br>
      rows of the distance matrix), <span class="math inline">\(d(i,j)\)</span> is the distance from node <span class="math inline">\(i\)</span><br>
      to node <span class="math inline">\(j\)</span>, <span class="math inline">\(R(i)\)</span> is the sum of the <span class="math inline">\(i\)</span>’th row, and <span class="math inline">\(C(j)\)</span> is the sum<br>
      of the <span class="math inline">\(j\)</span>’th column.</div>
<ol start="3" style="list-style-type: decimal">
<li>Identify the smallest pair <span class="math inline">\(f,g\)</span>
in the Q-matrix, and combine the corresponding nodes by adding an
intermediate node <span class="math inline">\(N_{int}\)</span> between
them in the tree.</li>
</ol>
<p><img src="nj2.png"></p>
<ol start="4" style="list-style-type: decimal">
<li>Calculate edge lengths from <span class="math inline">\(N_{int}\)</span> to <span class="math inline">\(f,g\)</span> with the following formula:</li>
</ol>
<p><span class="math display">\[\begin{align*}
d(N_{int}, f) &amp;= \frac{1}{2}(d[i,j]) + \frac{1}{2(n-2)}(R(i) -
C(j))\\
\\
d(N_{int}, g) &amp;= d[i,j] - d[N_{int}, f]
\end{align*}\]</span></p>
<div class="line-block">      In this case, we have <span class="math inline">\(d[1,2] = 2\)</span>, <span class="math inline">\(R(1) = 22\)</span>, <span class="math inline">\(C(2) = 20\)</span>. Thus, for our new node <span class="math inline">\(u\)</span>, we have:</div>
<span class="math display">\[\begin{align*}
d[u,1] &amp;= \frac{1}{2}(2) + \frac{1}{2(3)}(20-22) &amp;= \frac{2}{3}
\\
\\
d[u,2] &amp;= 2 - \frac{2}{3} &amp;= \frac{4}{3}
\end{align*}\]</span>
<ol start="5" style="list-style-type: decimal">
<li>Calculate distance from <span class="math inline">\(N_{int}\)</span>
to all other nodes <span class="math inline">\(h\)</span> with the
following formula:</li>
</ol>
<p><span class="math display">\[d(h, N_{int}) = \frac{1}{2}(d[h,g] +
d[h,f] - d[f,g])\]</span></p>
<ol start="6" style="list-style-type: decimal">
<li>Repeat until the full tree has been constructed.</li>
</ol>
<p><img src="nj3.png"></p>
<div class="line-block"> Once we only have three nodes left, we just
need to figure out the</div>
<p>branch lengths.</p>
<div class="line-block"> We can do this in the exact same way as
before.</div>
<p><img src="nj4.png"><img src="nj5.png"></p>
</div>
<div class="section level3">
<h3 id="r-implementation-1">R Implementation<a class="anchor" aria-label="anchor" href="#r-implementation-1"></a>
</h3>
<p>Neighbor-Joining trees can be created R with the following commands.
Note that the output is slightly different due to a different method for
calculating the initial distance matrix.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dend</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DECIPHER/man/TreeLine.html" class="external-link">TreeLine</a></span><span class="op">(</span>myDistMatrix<span class="op">=</span><span class="va">dm</span>, method<span class="op">=</span><span class="st">'NJ'</span><span class="op">)</span></span>
<span><span class="fu">plot_tree_unrooted</span><span class="op">(</span><span class="va">dend</span>, <span class="st">'Neighbor Joining'</span><span class="op">)</span></span></code></pre></div>
<p><img src="BuildingTrees_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="maximum-parsimony-trees">Maximum Parsimony Trees<a class="anchor" aria-label="anchor" href="#maximum-parsimony-trees"></a>
</h2>
<p>The last way to construct phylogenies that I will discuss is maximum
parsimony. This method relies on the assumption that the simplest method
is likely the best, and thus tries to find a tree that minimizes the
number of changes on the tree. Such a tree is called the most
“parsimonious” tree.</p>
<div class="section level3">
<h3 id="building-initial-tree">Building Initial Tree<a class="anchor" aria-label="anchor" href="#building-initial-tree"></a>
</h3>
<p>Let’s start by constructing a simple tree between our first two
sequences. The intermediate node is labeled with a consensus sequence,
which in this case is either sequence 1 or sequence 2. Both are equally
parsimonious, since they minimize the total number of transitions on
each branch.</p>
<p>I’ve labelled in pink the state at each node, and in orange the
number of transitions to go between the nodes connected by that
edge.</p>
<p><img src="mpn1.png"></p>
<p>We’ll now add each subsequent sequence. Let’s move on to sequence
3.</p>
<p>There are 3 total ways to add sequence 3 to our tree:</p>
<p><img src="mpn2.png"></p>
<p>Two of these trees place sequence 3 closer to sequence 1/2 than
sequence 2/1 (resp.), and the middle tree places all sequences equally
far apart.</p>
<p>Let’s now label these trees with intermediate states and see how
parsimonious they are:</p>
<p><img src="mpn3.png"></p>
<p>Short digression on how we find intermediate/ancestral sequences:</p>
<p>In the middle tree, there is only one choice for the ancestral
sequence, which is the one where each position is seen in the majority
of sequences. Each position in it (<span class="math inline">\(AGACT\)</span>) appears in <span class="math inline">\(2/3\)</span> of nodes connected to it (first
position of 1 and 2 is <span class="math inline">\(A\)</span>, second of
1 and 2 is <span class="math inline">\(G\)</span>, etc.). If there were
a position where all nodes were different, we would be free to pick any
of those bases (since all are equally represented and thus will be
equally parsimonious). There are often multiple ways to construct
ancestral states, but they should end up being equal. Look into Fitch
Parsimony and/or Sankoff Parsimony for a more rigorous definition of how
to find parsimonious ancestor states.</p>
<p>Back to the tutorial–the star tree is the most parsimonious! Our tree
now looks like this: <img src="mpn4.png"></p>
<p>Just to really drive the point home, let’s walk through adding
sequence 4 to the tree. There are now 4 options where we can put this
sequence: <img src="mpn5.png"></p>
<p>And again filling in ancestral states and finding transitions on each
branch:</p>
<p><img src="mpn6.png"></p>
<p>I’m going to skip walking through adding sequence 5 since it begins
to have a lot of possibilities, but suppose we arrive at the following
tree after adding sequence 5:</p>
<p><img src="mpn7.png"></p>
<p>Now we have our initial tree, which has a total parsimony cost of 7.
However, we’re not done yet. Adding branches sequentially doesn’t
guarantee we get an optimal solution, especially if we incorporate
optimizations to how we add branches. Many algorithms commonly only
minimize the parsimony cost of the branch at each addition (rather than
recalculating the entire tree) to speed up calculations. This can cause
us to miss better solutions due to local maxima. One common solution to
this is to perturb our initial tree.</p>
</div>
<div class="section level3">
<h3 id="nearest-neighbor-interchanges">Nearest Neighbor Interchanges<a class="anchor" aria-label="anchor" href="#nearest-neighbor-interchanges"></a>
</h3>
<p>The most common way to perturb a phylogenetic tree is with Nearest
Neighbor Interchanges (NNIs). For any particular set of 4 nodes, there
are 3 distinct ways to partition them into pairs. An NNI changes this
partition and rechecks the parsimony of the result to see if we can get
an improvement.</p>
<p>NNIs are significantly easier to explain with images. Suppose we have
a set of nodes <span class="math inline">\(\{a, b, c, d\}\)</span> such
that <span class="math inline">\(a,b\)</span> are separated from <span class="math inline">\(c,d\)</span>. Our possible NNIs then look like
this:</p>
<p><img src="nni1.png"></p>
<p>Let’s explore this on our initial tree by making an NNI on the edge
that separates <span class="math inline">\(\{4,5\}\)</span> from the
rest of the nodes. We’ll have two trees to check:</p>
<p><img src="nni2.png"></p>
<p>As mentioned above, the total cost for our current tree is 7. Do any
of these new trees do better?</p>
<p><img src="nni3.png"></p>
<p>So tree 2 is worse, but tree 1 is just as good as our initial tree!
I’m going to skip walking through all the other NNIs on other branches,
so let’s look at our final two equally parsimonious reconstructions
(node labels blue, branch lengths in orange):</p>
<p><img src="mptrees.png"></p>
<p>Note that we can remove branches with length 0.</p>
</div>
<div class="section level3">
<h3 id="r-implementation-2">R Implementation<a class="anchor" aria-label="anchor" href="#r-implementation-2"></a>
</h3>
<p>We can do maximum parsimony reconstructions in R using the following
code:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dend</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DECIPHER/man/TreeLine.html" class="external-link">TreeLine</a></span><span class="op">(</span><span class="va">sequenceSet</span>, method<span class="op">=</span><span class="st">'MP'</span><span class="op">)</span></span>
<span><span class="fu">plot_tree_unrooted</span><span class="op">(</span><span class="va">dend</span>, <span class="st">'Maximum Parsimony'</span><span class="op">)</span></span></code></pre></div>
<p><img src="BuildingTrees_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>Notice that this almost exactly matches one of our own
reconstructions! Parsimony reconstructions can differ based on how you
calculate parsimony, which can account for why the lengths in this plot
are slightly different than ours. Using a different transition cost
model (such as weighting transitions different than transversions) can
lead to different results.</p>
</div>
</div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Aidan Lakshman.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
